Similarity performance spell check

BrijeshKrishnan opened the pull request.

BrijeshKrishnan commented:
I did some quick change to similarity tool to make a performance improvement, Below is my finding
"
1. Execution time 6617.898900985718 Sec with simple work flow (main single thread)
2. Execution time 6119.476572036743 Sec with multiprocessing
3. Execution time 6082.6781940460205 Sec with multiprocessing using own queue
4. Execution time 412.2530097961426 Sec single thread with prepossessed data = > usage of numpy, sklearn
"
Added the change to the code to adopt the point number 4 mentioned above
Added spell check to the project as a CI line check
Moved from travis CI to github workflow
Added mutation testing config file
Code clean up

dsvellal commented on setup.py:
```diff
@@ -14,14 +29,15 @@
-)
\ No newline at end of file
+    python_requires='>=3.7',
+)
```
A new line here may also be helpful, to comply with standards: https://unix.stackexchange.com/a/18789

dsvellal commented on similarity_processor/similarity_cmd.py:
```diff
@@ -4,12 +4,12 @@
 sys.path.append(os.path.abspath(os.path.join
                                 (os.path.dirname(__file__), os.pardir
-                                 )))
+                                 )))# pragma: no mutate
```
Could you please justify why no mutate pragma was added. I am asking for justification only for record-keeping purposes!

dsvellal commented on similarity_processor/similarity_io.py:
```diff
@@ -1,15 +1,18 @@
-from similarity_processor.similarity_core import get_cosine
-from similarity_processor.similarity_core import text_to_vector
+import numpy as np
+from sklearn.feature_extraction.text import CountVectorizer # pylint: disable=E0401
+from sklearn.metrics.pairwise import cosine_similarity # pylint: disable=E0401
```
Please help justify why this pylint warning is disabled?

BrijeshKrishnan replied:
Removed in this instance. but it is present in verify_path.py and sanity_test.py. The lint disable is activated here since we are not packaging these folders and hence there are false positive lint errors so kept the suppression at local.

dsvellal commented on similarity_processor/similarity_io.py:
```diff
@@ -120,56 +132,53 @@ def __write_xlsx(self, data_f, name):
+        count_vect = CountVectorizer()
+        word_count_vector = count_vect.fit_transform(self.data_frame['Steps'].to_numpy())
+        c_sim = 100 * (cosine_similarity(word_count_vector))
+        self.data_frame["Potential Match"] = self.data_frame[self.uniq_header]
```
Nitpick - please see if this is valuable. 

Sometimes the key is referred with double quotes, and sometimes with single-quotes. Should we standardize it?

BrijeshKrishnan replied:
standardized it with " ", but in some cases this in not applicable if there are quote with in another.

dsvellal commented on test/test_functional.py:
```diff
@@ -2,13 +2,13 @@
 import unittest
-from tkinter import Tk
+# from tkinter import Tk
 import time
 import subprocess
 from test.test_resource import TestResource
 import pandas as pd
 from similarity_processor.similarity_io import SimilarityIO
-from similarity_processor.similarity_ui import TextSimilarityWindow
+# from similarity_processor.similarity_ui import TextSimilarityWindow
```
Should we remove the commented code?

BrijeshKrishnan replied:
They are UI test which can be done at desk, but CI it is not possible hence kept as commented. In travis there was a skip option. such option i am not able to get in github action. Hence keeping it commented untill i find a way to skip the test in CI

dsvellal commented on test/test_functional.py:
```diff
@@ -47,28 +47,28 @@ def test_below_ui(self):
-        time.sleep(10)
-        window.quit()
-        self.verify_functional_test(True)
+    # @unittest.skipIf("TRAVIS" in os.environ and os.environ["TRAVIS"] == "true", "Skipping this test on Travis CI.")
+    # def test_from_ui_new_text(self):
+    #     """Test function which injects the user input data at the presentation later
+    #     to check the end to end functionality"""
+    #     window = Tk()
+    #     win = TextSimilarityWindow(window)
+    #     win.check_is_new_text.invoke()
+    #     win.path_t.insert(0, str(TestResource.file_path))
+    #     win.uniq_id_t.insert(0, 0)
+    #     win.steps_t.insert(0, "1,2")
+    #     time.sleep(2)
+    #     win.new_text.insert(0, "a3 d4")
+    #     win.submit.invoke()
+    #     time.sleep(10)
+    #     window.quit()
+    #     self.verify_functional_test(True)
```
Should we remove the commented code?

BrijeshKrishnan replied:
They are UI test which can be done at desk, but CI it is not possible hence kept as commented. In travis there was a skip option. such option i am not able to get in github action. Hence keeping it commented untill i find a way to skip the test in CI

dsvellal commented on test_resource/cyclo_gate.py:
```diff
@@ -0,0 +1,129 @@
+"""Koninklijke Philips N.V., 2019 - 2020. All rights reserved."""
```
Thank you for putting this copyright header.

Should we add this to all the files if not present already?

BrijeshKrishnan replied:
Accepted. Added the header in all files

BrijeshKrishnan commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,62 @@
+        node-version: ${{ env.NODE_VERSION }}
+    - name: Install dependencies
+      run: |
+        python3 -m pip install --upgrade pip
```
Thanks for the comment. Indeed it will help to replicate same environment as CI I have abstracted the install, static analysis check, test etc in to python scripts. please have a review.

BrijeshKrishnan commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,62 @@
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+    - name: Aspell install
+      run: |
+        sudo apt-get update -qq
```
This is because of github action issue mentioned in https://github.com/actions/virtual-environments/issues/599
for Aspell it is also specified as the way to install and with out sudo it doesnt https://howtoinstall.co/en/ubuntu/trusty/aspell

BrijeshKrishnan commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,62 @@
+    - name: npm install
+      run: |
+        sudo npm install
+        sudo npm install -g jscpd@3.2.1
```
I have not faced and degradation . will keep an eye on it. if degrades will roll back to the version you mentioned and will give a try

BrijeshKrishnan commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,62 @@
+    - name: Static Analysis
+      run: |
+        pylint similarity_processor/ test/
+        jscpd --min-tokens 20 --reporters "json" --mode "strict" --format "python" -o . .
```
pylint , spelling, vulture is not having report and are gated at exit code Remaining are added as artifact

BrijeshKrishnan commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,62 @@
+  pull_request:
+    branches: [ master ]
+env:
+  NODE_VERSION: '10.x'                # set this to the node version to use
```
Added explicit version

dsvellal commented on build_scripts/dependencies_static_analysis_test_cov.py:
```diff
@@ -0,0 +1,126 @@
+    function check the repo for any yml linting errors
+    """
+    print("----TODO----- Markdown linting")
+    # call_subprocess("markdownlint *.md ")
+    # print("Stage linting md files -- COMPLETED & PASSED  --")
```
Work in progress?

BrijeshKrishnan replied:
yes, i would like to take this as part of separate checkin as it will call for a refactor of readme.md. I was able to get output locally on linting the md files. Just that i would like to refactor readme with needed more-info

dsvellal commented on build_scripts/performance_test.py:
```diff
@@ -0,0 +1,69 @@
+
+def validate_perfo(time_val, input_data):
+    """ Function to validate the performance of tool"""
+    if not time_val < 0.045 * int(input_data):
```
Would it be worth reading 0.045 from a configuration file instead of hard-coding it? That way, we can dynamically change the value if needed without modifying the code.

BrijeshKrishnan replied:
added it as command line input`python3 build_scripts/performance_test.py 0.045`

BrijeshKrishnan commented:
Addressed the suggestions and comments described by you. Thank you for the inputs.
1. I could not find a better way to ignore the lines (log, print) from mutation. If i come across something i will add to the project. 
2. I am yet to add the md files linting. Not to loose traction, i have added a "todo" in the script so that it will not be missed as pert of next checkin

Additional improvements made:
1. Added yml linting
2. added automatic deploy as part of work flow
3. Added automatic versioning for the tool based on git tag, so that need not have to modify the files
4. Added automatic installation of the newly created package in the build machine
5. Added sanity check on the newly created whl file
6. Added basic element of performance test to bench mark the speed of computation
7. Few of the build reports get published in the artifacts
8. Deploy to pypi as well as test pypi happens automatically

dsvellal commented:
Following up after our offline discussion! 

```
LGTM! Ship-it! 

             /|~~~   
           ///|      
         /////|      
       ///////|      
     /////////|      
   \==========|===/  
~~~~~~~~~~~~~~~~~~~~~
```
